File: webapps/frontend/app/viewmodels/containers/memcached.js
// Copyright (C) 2016 iNuron NV
//
// This file is part of Open vStorage Open Source Edition (OSE),
// as available from
//
//      http://www.openvstorage.org and
//      http://www.openvstorage.com.
//
// This file is free software; you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
// as published by the Free Software Foundation, in version 3 as it comes
// in the LICENSE.txt file of the Open vStorage OSE distribution.
//
// Open vStorage is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY of any kind.
/*global define */
define([
    'jquery', 'knockout',
    'ovs/generic', 'ovs/api'
], function($, ko, generic, api) {
    "use strict";
    return function() {
        var self = this;

        // Variables
        self.refreshHandle = undefined;
        self.refreshTimeout = undefined;

        // Obserables
        self.nodes = ko.observableArray([]);

        // Functions
        self.refresh = function() {
            if (generic.xhrCompleted(self.refreshHandle)) {
                self.refreshHandle = api.get('statistics/memcache')
                    .done(function(data) {
                        $.each(data.offline, function(index, node) {
                            var node_info, add = false;
                            $.each(self.nodes(), function(oindex, onode) {
                                if (onode.node() === node) {
                                    node_info = onode;
                                }
                            });
                            if (node_info === undefined) {
                                node_info = {
                                    node         : ko.observable(''),
                                    bytes        : ko.observable(''),
                                    currItems    : ko.observable(0),
                                    totalItems   : ko.observable(0),
                                    getHits      : ko.observable(0),
                                    cmdGet       : ko.observable(0),
                                    cmdGetps     : ko.deltaObservable(generic.formatShort),
                                    hitRate      : ko.observable(0),
                                    bytesRead    : ko.observable(''),
                                    bytesWritten : ko.observable(''),
                                    uptime       : ko.observable(0),
                                    raw          : ko.observable(''),
                                    online       : ko.observable(true)
                                };
                                node_info.cmdGetps(0);
                                add = true;
                            }
                            node_info.node(node);
                            node_info.online(false);
                            if (add) {
                                self.nodes.push(node_info);
                            }
                        });
                        $.each(data.nodes, function(index, node) {
                            var node_info, add = false, rawString = '', attribute;
                            $.each(self.nodes(), function(oindex, onode) {
                                if (onode.node() === node.node) {
                                    node_info = onode;
                                }
                            });
                            if (node_info === undefined) {
                                node_info = {
                                    node         : ko.observable(''),
                                    bytes        : ko.observable(''),
                                    currItems    : ko.observable(0),
                                    totalItems   : ko.observable(0),
                                    getHits      : ko.observable(0),
                                    cmdGet       : ko.observable(0),
                                    cmdGetps     : ko.deltaObservable(generic.formatShort),
                                    hitRate      : ko.observable(0),
                                    bytesRead    : ko.observable(''),
                                    bytesWritten : ko.observable(''),
                                    uptime       : ko.observable(0),
                                    raw          : ko.observable(''),
                                    online       : ko.observable(true)
                                };
                                node_info.cmdGetps(0);
                                add = true;
                            }
                            node_info.node(node.node);
                            node_info.online(true);
                            node_info.bytes(generic.formatBytes(node.bytes));
                            node_info.currItems(node.curr_items);
                            node_info.totalItems(node.total_items);
                            node_info.getHits(node.get_hits);
                            node_info.cmdGet(node.cmd_get);
                            node_info.cmdGetps(node.cmd_get);
                            node_info.hitRate(node.cmd_get === 0 ? 100 : generic.round(node.get_hits / node.cmd_get * 100, 2));
                            node_info.bytesRead(generic.formatBytes(node.bytes_read));
                            node_info.bytesWritten(generic.formatBytes(node.bytes_written));
                            node_info.uptime(node.uptime);

                            for (attribute in node) {
                                if (node.hasOwnProperty(attribute)) {
                                    rawString += generic.padRight(attribute, ' ', 25) + node[attribute].toString() + '\n';
                                }
                            }
                            node_info.raw(rawString);

                            if (add) {
                                self.nodes.push(node_info);
                            }
                        });
                    });
            }
        };
    };
});

File: webapps/api/backend/serializers/memcached.py
# Copyright (C) 2016 iNuron NV
#
# This file is part of Open vStorage Open Source Edition (OSE),
# as available from
#
#      http://www.openvstorage.org and
#      http://www.openvstorage.com.
#
# This file is free software; you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
# as published by the Free Software Foundation, in version 3 as it comes
# in the LICENSE.txt file of the Open vStorage OSE distribution.
#
# Open vStorage is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY of any kind.

"""
Memcache serializer module
"""
from rest_framework import serializers


class MemcacheSerializer(serializers.Serializer):
    """
    Serializes Memcache information
    """
    pid                  = serializers.Field()
    uptime               = serializers.Field()
    time                 = serializers.Field()
    version              = serializers.Field()
    rusage_user          = serializers.Field()
    rusage_system        = serializers.Field()
    curr_items           = serializers.Field()
    total_items          = serializers.Field()
    bytes                = serializers.Field()
    curr_connections     = serializers.Field()
    total_connections    = serializers.Field()
    cmd_get              = serializers.Field()
    cmd_set              = serializers.Field()
    get_hits             = serializers.Field()
    get_misses           = serializers.Field()
    evictions            = serializers.Field()
    bytes_read           = serializers.Field()
    bytes_written        = serializers.Field()
    limit_maxbytes       = serializers.Field()
    ovs_dal              = serializers.Field()

    class Meta:
        """
        Contains metadata regarding serializing fields
        """
        fields = ('pid', 'uptime', 'time', 'version', 'rusage_user', 'rusage_system', 'curr_items',
                  'total_items', 'bytes', 'curr_connections', 'total_connections', 'cmd_get',
                  'cmd_set', 'ovs_dal', 'get_hits', 'get_misses', 'evictions', 'bytes_read',
                  'bytes_written', 'limit_maxbytes')
